packages/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx

```diff
@@ -441,18 +509,62 @@ export const useGridRowReorder = (
         );
 
         if (validatedIndex !== -1) {
-          applyRowAnimation(() => {
-            apiRef.current.setRowIndex(dragRowId, validatedIndex);
-
-            // Emit the rowOrderChange event only once when the reordering stops.
-            const rowOrderChangeParams: GridRowOrderChangeParams = {
-              row: apiRef.current.getRow(dragRowId)!,
-              targetIndex: validatedIndex,
-              oldIndex: sourceRowIndex,
-            };
-
-            apiRef.current.publishEvent('rowOrderChange', rowOrderChangeParams);
-          });
+          try {
+            await applyRowAnimation(async () => {
+              // Wait for setRowIndex to complete (if it returns a Promise)
+              const result = apiRef.current.setRowIndex(dragRowId, validatedIndex);
+
+              // Handle both sync and async cases
+              if (result && typeof result.then === 'function') {
+                await result;
+              }
```

Comment:
Not sure if it's necessary (see https://github.com/mui/mui-x/pull/19401#discussion_r2403007128), but you can await function regardless whether it's sync or async:

```suggestion
              await apiRef.current.setRowIndex(dragRowId, validatedIndex);
```